### PR TITLE
Milestone 3 - Risk Management

### PR DESCRIPTION
**Pull request by FATCHIP-GmbH**

### CODE REVIEW

File: Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        parent::__construct($context, $data);
+        $this->elementFactory = $elementFactory;
+        $this->personStatus = $personStatus;
+        $this->creditScore = $creditScore;
+    }
+
+    /**
+     * Initialise form fields
+     *
+     * @return void
```
Comment (fjbender) [inline Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php]: Constructors shouldn't `@return` anything

File: Helper/Consumerscore.php [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     * @return int
+     */
+    public function getConsumerscoreSampleCounter()
+    {
+        $iCounter = $this->databaseHelper->getConfigParamWithoutCache(
+            self::CONFIG_KEY_CONSUMERSCORE_SAMPLE_COUNTER,
+            'creditrating',
+            'payone_protect'
+        );
+        if (empty($iCounter) or !is_numeric($iCounter)) {
```
Comment (fjbender) [inline Helper/Consumerscore.php]: Should use `||` here

File: Model/Api/Request/Addresscheck.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    public function __construct(
+        \Payone\Core\Helper\Shop $shopHelper,
+        \Payone\Core\Helper\Environment $environmentHelper,
+        \Payone\Core\Helper\Api $apiHelper,
+        \Payone\Core\Model\ResourceModel\ApiLog $apiLog,
+        \Payone\Core\Helper\Customer $customerHelper,
+        \Payone\Core\Model\ResourceModel\CheckedAddresses $addressesChecked
+    ) {
+        parent::__construct($shopHelper, $environmentHelper, $apiHelper, $apiLog, $customerHelper);
+        $this->addressesChecked = $addressesChecked;
```
Comment (fjbender) [inline Model/Api/Request/Addresscheck.php]: Property `$this->addressesChecked` is not defined/declared dynamically.

File: Model/Api/Request/Consumerscore.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Send request "addresscheck" to PAYONE server API
+     *
+     * @param  AddressInterface $oAddress
+     * @param  bool             $blIsBillingAddress
```
Comment (fjbender) [inline Model/Api/Request/Consumerscore.php]: This parameter doesn't exist.

File: Model/Plugins/BillingAddressManagement.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    protected function handleAddresscheck(AddressInterface $oAddress)
+    {
+        $aResponse = $this->addresscheck->getResponse($oAddress);
+        if (is_array($aResponse)) {
+            $sErrorMessage = false;
+            if ($aResponse['status'] == 'INVALID') {
+                $sErrorMessage = $this->addresscheck->getInvalidMessage($aResponse['customermessage']);
+            } elseif ($aResponse['status'] == 'ERROR') {
+                if ($this->toolkitHelper->getConfigParam('handle_response_error', 'address_check', 'payone_protect') == 'stop_checkout') {
```
Comment (fjbender) [inline Model/Plugins/BillingAddressManagement.php]: `$this->toolkitHelper` is not defined.

File: Model/Plugins/ShippingAddressManagement.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    protected function handleAddresscheck(AddressInterface $oAddress)
+    {
+        $aResponse = $this->addresscheck->getResponse($oAddress);
+        if (is_array($aResponse)) {
+            $sErrorMessage = false;
+            if ($aResponse['status'] == 'INVALID') {
+                $sErrorMessage = $this->addresscheck->getInvalidMessage($aResponse['customermessage']);
+            } elseif ($aResponse['status'] == 'ERROR') {
+                if ($this->toolkitHelper->getConfigParam('handle_response_error', 'address_check', 'payone_protect') == 'stop_checkout') {
+                    $sErrorMessage = $this->toolkitHelper->getConfigParam('stop_checkout_message', 'address_check', 'payone_protect');
```
Comment (fjbender) [inline Model/Plugins/ShippingAddressManagement.php]: `$this->toolkitHelper` is not defined.

File: Model/Plugins/MethodList.php [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        return $this->checkoutSession->getQuote();
+    }
+
+    /**
+     * Determine if creditrating is needed
+     *
+     * @return bool
+     */
+    protected function isCreditratingNeeded()
```
Comment (fjbender) [inline Model/Plugins/MethodList.php]: This Method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CModel%5CPlugins%5CMethodList%3A%3AisCreditratingNeeded

File: Model/WebApi/Addresscheck.php [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+    /**
+     * Send addresscheck request and handle the response object
+     *
+     * @param  AddresscheckResponse                     $oResponse
+     * @param  \Magento\Quote\Api\Data\AddressInterface $oAddress
+     * @param  bool                                     $blIsBillingAddress
+     * @return AddresscheckResponse
+     */
+    protected function handleAddresscheck(
```
Comment (fjbender) [inline Model/WebApi/Addresscheck.php]: This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CModel%5CWebApi%5CAddresscheck%3A%3AhandleAddresscheck

File: Model/Risk/Addresscheck.php [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        return $this->addressCorrected;
+    }
+
+    /**
+     * Change the address according to the response
+     *
+     * @param  AddressInterface $oAddress
+     * @return AddressInterface
+     */
+    public function correctAddress(AddressInterface $oAddress)
```
Comment (fjbender) [inline Model/Risk/Addresscheck.php]: This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CModel%5CRisk%5CAddresscheck%3A%3AcorrectAddress

File: Observer/CheckoutSubmitBefore.php [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        return $this->consumerscoreHelper->getConfigParam($sParam, $sGroup, 'payone_protect');
+    }
+
+    /**
+     * Determine if creditrating is needed
+     *
+     * @param  Quote $oQuote
+     * @return bool
+     */
+    protected function isCreditratingNeeded(Quote $oQuote)
```
Comment (fjbender) [inline Observer/CheckoutSubmitBefore.php]: This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CObserver%5CCheckoutSubmitBefore%3A%3AisCreditratingNeeded

File: Observer/CheckoutSubmitBefore.php [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        return false;
+    }
+
+    /**
+     * Filter payment methods by the creditrating result if applicable
+     *
+     * @param  AddressInterface $oBilling
+     * @return void
+     */
+    protected function getScoreByCreditrating(AddressInterface $oBilling)
```
Comment (fjbender) [inline Observer/CheckoutSubmitBefore.php]: This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CObserver%5CCheckoutSubmitBefore%3A%3AgetScoreByCreditrating

File: Model/Plugins/MethodList.php [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    /**
+     * Filter methods by the worst score
+     *
+     * @param  MethodInterface[] $aPaymentMethods
+     * @param  string            $sWorstScore
+     * @return MethodInterface[]
+     */
+    protected function filterMethodsByScore($aPaymentMethods, $sWorstScore)
```
Comment (fjbender) [inline Model/Plugins/MethodList.php]: This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CModel%5CPlugins%5CMethodList%3A%3AfilterMethodsByScore

File: Model/Plugins/ShippingAddressManagement.php [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        $this->addresscheck = $addresscheck;
+    }
+
+    /**
+     * Execute addresscheck and return the response
+     *
+     * @param  AddressInterface $oAddress
+     * @return array
+     */
+    protected function handleAddresscheck(AddressInterface $oAddress)
```
Comment (fjbender) [inline Model/Plugins/ShippingAddressManagement.php]: This is somewhat duplicated for Shipping and Billing addresses. Can we avoid the duplication?

File: Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php (hunk 1 above)
Comment (Fatchip-Support) [inline Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php]: Dies ist kein php Konstruktor "__construct()", sondern eine Magento 2 Core Methode namens "_construct()".
Diese Methode hat auch in allen Anwendungen im Magento 2 Core ein "@return void" stehen.

File: Helper/Consumerscore.php (hunk 2 above)
Comment (Fatchip-Support) [inline Helper/Consumerscore.php]: done - was old Magento 1 code

File: Model/Api/Request/Addresscheck.php (hunk 3 above)
Comment (Fatchip-Support) [inline Model/Api/Request/Addresscheck.php]: done

File: Model/Api/Request/Consumerscore.php (hunk 4 above)
Comment (Fatchip-Support) [inline Model/Api/Request/Consumerscore.php]: done

File: Model/Plugins/BillingAddressManagement.php (hunk 5 above)
Comment (Fatchip-Support) [inline Model/Plugins/BillingAddressManagement.php]: done

File: Model/Plugins/ShippingAddressManagement.php (hunk 13 above)
Comment (Fatchip-Support) [inline Model/Plugins/ShippingAddressManagement.php]: done

File: Model/Plugins/ShippingAddressManagement.php (hunk 6 above)
Comment (Fatchip-Support) [inline Model/Plugins/ShippingAddressManagement.php]: done

File: Model/Plugins/MethodList.php (hunk 12 above)
Comment (Fatchip-Support) [inline Model/Plugins/MethodList.php]: in 2017

File: Model/Plugins/MethodList.php (hunk 7 above)
Comment (Fatchip-Support) [inline Model/Plugins/MethodList.php]: in 2017

File: Model/WebApi/Addresscheck.php (hunk 8 above)
Comment (Fatchip-Support) [inline Model/WebApi/Addresscheck.php]: in 2017

File: Model/Risk/Addresscheck.php (hunk 9 above)
Comment (Fatchip-Support) [inline Model/Risk/Addresscheck.php]: in  2017

File: Observer/CheckoutSubmitBefore.php (hunk 10 above)
Comment (Fatchip-Support) [inline Observer/CheckoutSubmitBefore.php]: in 2017

File: Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php (hunk 1 above)
Comment (fjbender) [inline Block/Adminhtml/Config/Form/Field/PersonStatusMapping.php]: Okay

File: Model/WebApi/Addresscheck.php [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+     * Handle the response according to its return status
+     *
+     * @param  AddresscheckResponse $oResponse
+     * @param  array                $aResponse
+     * @return AddresscheckResponse
+     */
+    protected function handleResponse(AddresscheckResponse $oResponse, $aResponse)
+    {
+        if ($aResponse['status'] == 'VALID') { // data was checked successfully
+            $oAddress = $this->addresscheck->correctAddress($oAddress);
```
Comment (fjbender) [inline Model/WebApi/Addresscheck.php]: Bug here, `$oAddress` is not defined: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/93ca0b97-237d-4813-bf38-3c1592c5d997/issues/files/Model/WebApi/Addresscheck.php?status=new&selectedSeverities%5B0%5D=10&orderField=path&order=asc&honorSelectedPaths=0

File: Block/Adminhtml/Information.php [hunk 15]
```diff
@@ -57,6 +57,6 @@ public function getHeader()
      */
     public function getPayoneUrl()
     {
-        return '//www.payone.de/embedded-sites/magento/information/';
+        return 'http://www.payone.de/embedded-sites/magento/information/';
```
Comment (fjbender) [inline Block/Adminhtml/Information.php]: This seems like a revert of a previous commit.

File: Controller/Onepage/Cancel.php [hunk 16]
```diff
@@ -67,7 +68,7 @@ public function __construct(
     /**
      * Checkout is canceled and old basket is reactivated
      *
-     * @return \Magento\Framework\Controller\Result\Redirect
```
Comment (fjbender) [inline Controller/Onepage/Cancel.php]: This seems like a revert of a previous commit.

File: Model/Methods/BaseMethod.php [hunk 17]
```diff
@@ -85,21 +85,6 @@
...
-    protected $_canCapturePartial = true;
-
-    /**
-     * Determines if payment type can use partial refunds
-     *
-     * @var bool
-     */
-    protected $_canRefundInvoicePartial = true;
-
-    /**
```
Comment (fjbender) [inline Model/Methods/BaseMethod.php]: This seems like a revert of a previous commit.

File: Model/Plugins/BillingAddressManagement.php [hunk 18]
```diff
@@ -28,7 +28,6 @@
 
 use Magento\Quote\Api\Data\AddressInterface;
 use Magento\Quote\Model\BillingAddressManagement as BillingAddressManagementOrig;
-use Magento\Framework\Exception\LocalizedException;
```
Comment (fjbender) [inline Model/Plugins/BillingAddressManagement.php]: This seems like a revert of a previous commit.

Comment (FATCHIP-GmbH) [inline Model/Plugins/BillingAddressManagement.php]: Nope, not a revert just a change we made

**Pull request merged**